cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/BrokerSetResolutionHelper.java

```diff
@@ -5,73 +5,96 @@
+  private void initializeBrokerSetMappings(Map<Integer, String> rackIdByBrokerId, BrokerSetResolver brokerSetResolver)
       throws BrokerSetResolutionException {
-    _brokersByBrokerSetId = brokerSetResolver.brokerIdsByBrokerSetId(clusterModel)
-                                             .entrySet()
-                                             .stream()
-                                             .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue()
-                                                                                                .stream()
-                                                                                                .map(id -> clusterModel.broker(id))
-                                                                                                .filter(Objects::nonNull)
-                                                                                                .collect(Collectors.toSet())));
-
+    _brokersByBrokerSetId = brokerSetResolver.brokerIdsByBrokerSetId(rackIdByBrokerId);
     _brokerSetIdByBrokerId = new HashMap<>();
 
     _brokersByBrokerSetId.forEach((brokerSet, brokers) -> {
-      brokers.forEach(broker -> _brokerSetIdByBrokerId.put(broker.id(), brokerSet));
+      brokers.forEach(broker -> _brokerSetIdByBrokerId.put(broker, brokerSet));
     });
   }
 
   /**
-   * Helper function to get Broker to BrokerSet mapping
+   * Helper function to get Broker by BrokerSet mapping
    *
    * @return broker by broker set id Mapping
    */
-  public Map<String, Set<Broker>> brokersByBrokerSetId() {
+  public Map<String, Set<Integer>> brokersByBrokerSetId() {
     return _brokersByBrokerSetId;
   }
 
+  /**
+   * Helper function to get BrokerSet by Broker mapping
+   *
+   * @return broker set id by broker id Mapping
+   */
+  public Map<Integer, String> brokerSetIdByBrokerId() {
+    return _brokerSetIdByBrokerId;
+  }
+
   /**
    * Helper function to get brokerSet for a broker
    *
-   * @param broker broker object
+   * @param brokerId broker object
    * @return broker set id
    */
-  public String brokerSetId(Broker broker) throws BrokerSetResolutionException {
-    String brokerSetId = _brokerSetIdByBrokerId.get(broker.id());
+  public String brokerSetId(Integer brokerId) throws BrokerSetResolutionException {
+    String brokerSetId = _brokerSetIdByBrokerId.get(brokerId);
 
     if (brokerSetId != null) {
       return brokerSetId;
     }
 
-    throw new BrokerSetResolutionException(String.format("Failed to resolve BrokerSet for Broker %s", broker));
+    throw new BrokerSetResolutionException(String.format("Failed to resolve BrokerSet for Broker %s", brokerId));
+  }
+
+  /**
+   * Get the broker to rack mapping from cluster
+   * @param cluster the kafka cluster
+   * @return the map of broker id to rack id
+   */
+  public static Map<Integer, String> getRackIdByBrokerIdMapping(Cluster cluster) {
```

Comment:
Does this need to be public? Same goes for the other method with the same name.

Reply:
Actually, they are also used in unit tests. So I'd keep them as public static.